dm/pkg/checker/privilege.go

```diff
@@ -122,89 +143,215 @@ func (pc *SourceReplicatePrivilegeChecker) Name() string {
+				// from https://dev.mysql.com/doc/refman/5.7/en/privileges-provided.html#priv_all
+				if privElem.Priv == mysql.AllPriv {
+					for priv := range lackPriv {
+						if priv == mysql.GrantPriv {
+							continue
+						}
+						if _, ok := lackPriv[priv][dbName]; !ok {
+							continue
+						}
+						delete(lackPriv[priv], dbName)
+						if len(lackPriv[priv]) == 0 {
+							delete(lackPriv, priv)
+						}
 					}
-				} else {
-					// currently, only SELECT privilege goes here. we didn't require SELECT to be granted globally,
-					// dumpling could report error if an allow-list table is lack of privilege.
-					// we only check that SELECT is granted on all columns, otherwise we can't SHOW CREATE TABLE
-					if len(privElem.Cols) == 0 {
-						delete(lackGrants, privElem.Priv)
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv]; !ok {
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv][dbName]; !ok {
+					continue
+				}
+				// dumpling could report error if an allow-list table is lack of privilege.
+				// we only check that SELECT is granted on all columns, otherwise we can't SHOW CREATE TABLE
+				if privElem.Priv == mysql.SelectPriv && len(privElem.Cols) != 0 {
+					continue
+				}
+				delete(lackPriv[privElem.Priv], dbName)
+				if len(lackPriv[privElem.Priv]) == 0 {
+					delete(lackPriv, privElem.Priv)
+				}
+			}
+		case ast.GrantLevelTable:
+			for _, privElem := range grantStmt.Privs {
+				// all privileges available at a given privilege level (except GRANT OPTION)
+				// from https://dev.mysql.com/doc/refman/5.7/en/privileges-provided.html#priv_all
+				if privElem.Priv == mysql.AllPriv {
+					for priv := range lackPriv {
+						if priv == mysql.GrantPriv {
+							continue
+						}
+						if _, ok := lackPriv[priv][dbName][tableName]; !ok {
+							continue
+						}
+						delete(lackPriv[priv][dbName], tableName)
+						if len(lackPriv[priv][dbName]) == 0 {
+							delete(lackPriv[priv], dbName)
+						}
+						if len(lackPriv[priv]) == 0 {
+							delete(lackPriv, priv)
+						}
 					}
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv]; !ok {
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv][dbName]; !ok {
+					continue
+				}
+				if _, ok := lackPriv[privElem.Priv][dbName][tableName]; !ok {
+					continue
+				}
+				// dumpling could report error if an allow-list table is lack of privilege.
+				// we only check that SELECT is granted on all columns, otherwise we can't SHOW CREATE TABLE
+				if privElem.Priv == mysql.SelectPriv && len(privElem.Cols) != 0 {
+					continue
+				}
+				delete(lackPriv[privElem.Priv][dbName], tableName)
+				if len(lackPriv[privElem.Priv][dbName]) == 0 {
+					delete(lackPriv[privElem.Priv], dbName)
+				}
+				if len(lackPriv[privElem.Priv]) == 0 {
+					delete(lackPriv, privElem.Priv)
+				}
+			}
+		}
+	}
+
+	if len(lackPriv) == 0 {
+		return nil
+	}
+	var buffer bytes.Buffer
```

Comment:
```suggestion
	var buffer strings.Builder
```
use string builder should be a more ergonomic way here.